In [2]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind

In [3]:
with open('ids_first_company_positive.txt', 'r') as file:
     content = file.read()


In [4]:
import re
cleaned_text3 = re.sub(r'[a-zA-Zа-яА-Я]', '', content)

In [5]:
delimiters = (",", ";", "\n", " ")

In [6]:
result3 = re.split('|'.join(map(re.escape, delimiters)), cleaned_text3)

In [7]:
df7 = pd.DataFrame(result3[5:-2])

df7 = df7.rename(columns={0:'id'})[['id']]

df7['id'] = pd.to_numeric(df7['id'], errors='coerce')


df7 = df7.dropna(subset=['id'])

print(df7)



            id
0      88128.0
1     171960.0
2     142117.0
3     114589.0
4     146123.0
...        ...
5026   20387.0
5027   61130.0
5028   18359.0
5029   87690.0
5030   81928.0

[5022 rows x 1 columns]


In [8]:
value_to_remove = 0

df7 = df7[(df7['id'] != value_to_remove)]

In [9]:
df_positive = df7
df_positive

id
0      88128.0
1     171960.0
2     142117.0
3     114589.0
4     146123.0
...        ...
5026   20387.0
5027   61130.0
5028   18359.0
5029   87690.0
5030   81928.0

[5020 rows x 1 columns]

In [10]:
with open('ids_first_company_negative.csv', 'r') as file:
     content = file.read()


In [11]:
import re
cleaned_text3 = re.sub(r'[a-zA-Zа-яА-Я]', '', content)

In [12]:
delimiters = (",", ";", "\n", " ")

In [13]:
result3 = re.split('|'.join(map(re.escape, delimiters)), cleaned_text3)

In [14]:
df7 = pd.DataFrame(result3[5:-2])

df7 = df7.rename(columns={0:'id'})[['id']]

df7['id'] = pd.to_numeric(df7['id'], errors='coerce')


df7 = df7.dropna(subset=['id'])

print(df7)

            id
0      82450.0
1      63400.0
2       7138.0
3     102118.0
4     153511.0
...        ...
6040    6809.0
6041   89121.0
6042  161445.0
6043  162570.0
6044   92547.0

[4958 rows x 1 columns]


In [15]:
value_to_remove = 0

df7 = df7[(df7['id'] != value_to_remove)]

In [16]:
df_negative = df7
df_negative

id
0      82450.0
1      63400.0
2       7138.0
3     102118.0
4     153511.0
...        ...
6040    6809.0
6041   89121.0
6042  161445.0
6043  162570.0
6044   92547.0

[4958 rows x 1 columns]

In [17]:
df_negative['id']=df_negative['id'].astype(int)

In [18]:
df_positive['id']=df_positive['id'].astype(int)

In [19]:
df = pd.read_pickle(' all_data.pkl ')

In [20]:
df

id  gender  age education  city  country  lbt_coef   ac_coef  \
0           0       0   36   среднее  1201       32  5.078678 -0.307147   
1           0       0   36   среднее  1201       32  5.078678 -0.307147   
2           0       0   36   среднее  1201       32  5.078678 -0.307147   
3           4       0   35   среднее  1134       32  4.569378  0.063693   
4           4       0   35   среднее  1134       32  4.569378  0.063693   
...       ...     ...  ...       ...   ...      ...       ...       ...   
786255  15885       1   33   среднее  1190       32  7.321749  0.470877   
786256  15885       1   33   среднее  1190       32  7.321749  0.470877   
786257  15885       1   33   среднее  1190       32  7.321749  0.470877   
786258  15885       1   33   среднее  1190       32  7.321749  0.470877   
786259  15885       1   33   среднее  1190       32  7.321749  0.470877   

         sm_coef  personal_coef                                      product  \
0       0.959027         0.5072  Велосипед горный женский Stern Mira 2.0 26"   
1       0.959027         0.5072                              Стол Outventure   
2       0.959027         0.5072             Набор Outventure: стол + 4 стула   
3       0.820892         0.5072   Роликовые коньки детские Roces Orlando III   
4       0.820892         0.5072                            Шлем FILA NRK FUN   
...          ...            ...                                          ...   
786255  0.699928         0.4304                      Поло мужское Outventure   
786256  0.699928         0.4304        Майка мужская Babolat Complete Muscle   
786257  0.699928         0.4304              Майка мужская Protest Prtihiaka   
786258  0.699928         0.4304                               Массажер Demix   
786259  0.699928         0.4304             Тренажер гироскопический KETTLER   

                 colour   cost  product_sex  base_sale  dt  
0           белый/синий  13599          0.0          1   7  
1               зелёный   1499          0.0          0  37  
2               бежевый   4799          0.0          0  37  
3        черный/розовый   9199          0.0          0   0  
4                 белый   4999          0.0          0   0  
...                 ...    ...          ...        ...  ..  
786255            синий   1599          1.0          0  24  
786256  оранжевый/синий   3899          1.0          0  24  
786257      белый/принт   1679          1.0          1  49  
786258    темно-зелёный    399          1.0          0  52  
786259          голубой   1799          1.0          0  52  

[786260 rows x 16 columns]

In [21]:
df_com1 = df[(df['dt'] >= 5) & (df['dt'] <= 16)]

In [22]:
df_com1

id  gender  age education  city  country  lbt_coef   ac_coef  \
0            0       0   36   среднее  1201       32  5.078678 -0.307147   
12           6       1   52   среднее  1188       32  8.150379  0.075200   
13           6       1   52   среднее  1188       32  8.150379  0.075200   
14           6       1   52   среднее  1188       32  8.150379  0.075200   
15           6       1   52   среднее  1188       32  8.150379  0.075200   
...        ...     ...  ...       ...   ...      ...       ...       ...   
786228  135183       1   44   среднее  1134       32  7.585086 -0.077493   
786233  125993       1   46   среднее  1134       32  8.063351 -0.105398   
786238  167019       0   71   среднее  1134       32  6.172481  0.354901   
786243  116355       1   32   среднее  1188       32  7.084163 -0.132395   
786252   15885       1   33   среднее  1190       32  7.321749  0.470877   

         sm_coef  personal_coef  \
0       0.959027         0.5072   
12      0.836140         0.4304   
13      0.836140         0.4304   
14      0.836140         0.4304   
15      0.836140         0.4304   
...          ...            ...   
786228  0.672522         0.4304   
786233  0.710718         0.4304   
786238  1.491271         0.5072   
786243  0.509138         0.4304   
786252  0.699928         0.4304   

                                                  product         colour  \
0             Велосипед горный женский Stern Mira 2.0 26"    белый/синий   
12                  Футболка мужская Kappa Authentic Core    темно-синий   
13                  Бейсболка детская Nike Futura Mash Up          синий   
14                            Носки детские Demix, 2 пары    синий/серый   
15            Кроссовки мужские adidas Lite Racer CLN 2.0         черный   
...                                                   ...            ...   
786228              Ботинки женские Scarpa Boreas GTX Wmn  голубой/серый   
786233  Сандалии женские Merrell Terran 3 Cush Convert...   черный/серый   
786238          Рубашка с коротким рукавом женская Termit  синий/зеленый   
786243           Велосипед горный Trek Marlin 6 29", 2022         черный   
786252                        Кроссовки мужские FILA Base    темно-синий   

         cost  product_sex  base_sale  dt  
0       13599          0.0          1   7  
12        999          1.0          0   8  
13       1119          1.0          1  11  
14        299          1.0          0  11  
15       7399          1.0          0  11  
...       ...          ...        ...  ..  
786228  18199          0.0          0  13  
786233   6999          0.0          0   7  
786238   1999          0.0          0   6  
786243  69999          1.0          0  10  
786252   3999          1.0          0   7  

[229265 rows x 16 columns]

In [23]:
negative = df_negative.to_dict()
positive = df_positive.to_dict()


In [24]:
df_com1_pos = pd.merge(df_com1,df_positive,left_on = 'id',right_on = 'id',how='inner')

In [25]:
df_com1_neg = pd.merge(df_com1,df_negative,left_on = 'id',right_on = 'id',how='inner')

In [26]:
df_com1_pos

id  gender  age education  city  country  lbt_coef   ac_coef  \
0           6       1   52   среднее  1188       32  8.150379  0.075200   
1           6       1   52   среднее  1188       32  8.150379  0.075200   
2           6       1   52   среднее  1188       32  8.150379  0.075200   
3           6       1   52   среднее  1188       32  8.150379  0.075200   
4           6       1   52   среднее  1188       32  8.150379  0.075200   
...       ...     ...  ...       ...   ...      ...       ...       ...   
25116  105219       1   46    высшее  1134       32  8.009999 -0.373584   
25117  105219       1   46    высшее  1134       32  8.009999 -0.373584   
25118  105219       1   46    высшее  1134       32  8.009999 -0.373584   
25119  105219       1   46    высшее  1134       32  8.009999 -0.373584   
25120   15885       1   33   среднее  1190       32  7.321749  0.470877   

        sm_coef  personal_coef                                      product  \
0      0.836140         0.4304        Футболка мужская Kappa Authentic Core   
1      0.836140         0.4304        Бейсболка детская Nike Futura Mash Up   
2      0.836140         0.4304                  Носки детские Demix, 2 пары   
3      0.836140         0.4304  Кроссовки мужские adidas Lite Racer CLN 2.0   
4      0.836140         0.4304  Кроссовки мужские adidas Lite Racer CLN 2.0   
...         ...            ...                                          ...   
25116  1.401091         0.4688                         Носки adidas, 3 пары   
25117  1.401091         0.4688                    Солнцезащитные очки Demix   
25118  1.401091         0.4688                                   Лапа Demix   
25119  1.401091         0.4688                        Чехол для очков Kappa   
25120  0.699928         0.4304                  Кроссовки мужские FILA Base   

            colour  cost  product_sex  base_sale  dt  
0      темно-синий   999          1.0          0   8  
1            синий  1119          1.0          1  11  
2      синий/серый   299          1.0          0  11  
3           черный  7399          1.0          0  11  
4           черный  7399          1.0          0  13  
...            ...   ...          ...        ...  ..  
25116       черный  1199          1.0          0   7  
25117   мультицвет  1189          1.0          1   7  
25118       черный  1299          1.0          0   7  
25119        синий   199          1.0          0   7  
25120  темно-синий  3999          1.0          0   7  

[25121 rows x 16 columns]

In [27]:
df_com1_neg

id  gender  age education  city  country  lbt_coef   ac_coef  \
0          23       1   44   среднее  1178       32  8.122966  0.442953   
1          23       1   44   среднее  1178       32  8.122966  0.442953   
2          70       0   31    высшее  1185       32  5.137719  0.574183   
3          70       0   31    высшее  1185       32  5.137719  0.574183   
4         110       0   48    высшее   338       38  5.645520 -1.278326   
...       ...     ...  ...       ...   ...      ...       ...       ...   
19876  121124       1   38   среднее  1134       32  4.857486 -0.227271   
19877  121124       1   38   среднее  1134       32  4.857486 -0.227271   
19878   11558       1   33   среднее  1185       32  5.206464 -0.124576   
19879   11558       1   33   среднее  1185       32  5.206464 -0.124576   
19880  170390       1   39   среднее  1188       32  7.996322  0.356278   

        sm_coef  personal_coef  \
0      0.560832         0.4304   
1      0.560832         0.4304   
2      1.525770         0.5584   
3      1.525770         0.5584   
4      1.202653         0.3792   
...         ...            ...   
19876  1.132259         0.5072   
19877  1.132259         0.5072   
19878  0.724404         0.5072   
19879  0.724404         0.5072   
19880  0.932371         0.4304   

                                                 product             colour  \
0                                    Шорты женские Demix             черный   
1                                       Лиф женский Joss         коралловый   
2                   Сандалии женские Outventure Athens W         графитовый   
3                       Кроссовки мужские FILA Chain Smr               хаки   
4                   Сандалии женские Northland Wandern W             чёрный   
...                                                  ...                ...   
19876   Полуботинки женские Skechers Seager Minor League              серый   
19877  Ботинки утепленные женские Skechers On-The-Go Joy  светло-коричневый   
19878                                       Рюкзак Demix      розовый/серый   
19879                             Рюкзак DC Shoes Spassi      синий/розовый   
19880                        Тент Outventure Royal House            бежевый   

        cost  product_sex  base_sale  dt  
0       1999          0.0          0  16  
1        649          0.0          1  16  
2       3999          0.0          0   5  
3       5999          1.0          0  16  
4       3999          0.0          0   5  
...      ...          ...        ...  ..  
19876   4899          0.0          1  16  
19877   5319          0.0          1  16  
19878   1049          1.0          1  15  
19879   1699          1.0          1  16  
19880  18999          1.0          0  16  

[19881 rows x 16 columns]

In [28]:
test_cash = df_com1_pos['cost'].to_list()
control_cash = df_com1_neg['cost'].to_list()

In [29]:
test = np.array(test_cash)
control = np.array(control_cash)

In [30]:
print(ttest_ind(test_cash,control_cash,equal_var=False))
print('средняя выручка тестовой группы=',test.mean())
print('средняя выручка контрольной группы=',control.mean())

TtestResult(statistic=-2.9039660061680435, pvalue=0.0036866423022580723, df=41112.42889725759)
средняя выручка тестовой группы= 5312.669638947494
средняя выручка контрольной группы= 5570.422664855892


In [31]:
print('общая выручка тестовой группы=',test.sum())
print('общая выручка контрольной группы=',control.sum())

общая выручка тестовой группы= 133459574
общая выручка контрольной группы= 110745573


### Бизнес-рекомендация: Были рассмотренны две группы,для которых была предложенна скидка и для которых не было предложенно скидки,проверка результатов эксперимента статистическим критерием и анализом выручки показала,что контрольная группа(без скидки) превосходит тестовую по pvalue,так же средняя выручка контрольной группы выше,но общая выручка больше у тестовой группы и тестовая группа больше контрольной на 5000 покупателей,соответсвенно метрика средней выручки ниже в тестовой группе,но метрика количества покупок и общей выручке выше,этот эксперимент можно разворачивать(применять ко всем пользователям),но стоит увеличить метрику средней выручки.

In [88]:

dfcom2 = df[(df['city'] == 1134) & (df['dt'].isin([15, 45]))] 
dfcom2

id  gender  age education  city  country  lbt_coef   ac_coef  \
360         98       0   31   среднее  1134       32  4.967616 -0.151891   
376        101       0   33   среднее  1134       32  4.555538  0.122977   
410        105       1   33   среднее  1134       32  7.600942 -0.186262   
481        130       0   63   среднее  1134       32  5.284831  0.214128   
505        148       0   40   среднее  1134       32  4.789172  0.165627   
...        ...     ...  ...       ...   ...      ...       ...       ...   
786161   97334       1   32   среднее  1134       32  4.927617 -0.282740   
786170    3646       0   58   среднее  1134       32  7.802182  0.165915   
786171    3646       0   58   среднее  1134       32  7.802182  0.165915   
786229  135183       1   44   среднее  1134       32  7.585086 -0.077493   
786230  135183       1   44   среднее  1134       32  7.585086 -0.077493   

         sm_coef  personal_coef                                     product  \
360     1.354463         0.5072               SUP Easy Rider WAVE 10'6 Haki   
376     0.641520         0.5072         Джемпер флисовый женский Outventure   
410     0.988834         0.4304         Электросамокат HIPER Model A2 Black   
481     1.501723         0.5072  Ботинки утепленные женские Geox Braies ABX   
505     1.419092         0.5072                Бейсболка детская Outventure   
...          ...            ...                                         ...   
786161  0.891139         0.5072                    Шорты женские PUMA Rebel   
786170  1.195541         0.4304                         Лента силовая Demix   
786171  1.195541         0.4304               Эспандер универсальный Torneo   
786229  0.672522         0.4304             Бриджи женские Salomon Wayfarer   
786230  0.672522         0.4304             Шорты женские Protest Prtlearie   

                colour   cost  product_sex  base_sale  dt  
360            зеленый  28500          0.0          0  45  
376           небесный   4499          0.0          0  15  
410            Unknown  29999          1.0          0  15  
481              серый   5999          0.0          1  15  
505           милитари    799          0.0          0  15  
...                ...    ...          ...        ...  ..  
786161         Unknown   1749          0.0          0  15  
786170      фиолетовый    999          0.0          0  15  
786171  зеленый/черный    999          0.0          0  15  
786229          черный   4899          0.0          1  45  
786230          черный   2589          0.0          1  45  

[10864 rows x 16 columns]

In [92]:
com2cash = dfcom2['cost']

In [93]:
com2cash = np.array(com2cash)

In [94]:
com2cash.mean()

5800.096373343152

### Выводы по второму тестированию: Выборка состоит из 10864 покупателей,средняя выручка по всей выборке 5800,нет данных для сравнения,поэтому нужно сравнить это с выручкой за предыдущие схожие сроки и сделать выводы,в сравнении с выручкой по первому эксперименту,это хороший результат,он выше,чем в первом эксперименте,но по этому сравнению сделать выводы нельзя